### PR TITLE
samples: mpsl: timeslot: fix timeslot sample

### PR DESCRIPTION
**Pull request by laxiLang**
Calling SWI rather than calling k_msgq_put in
Zero Latency IRQ (mpsl timeslot callback), which
is not allowed to call kernel API. Fix the same
issue in dm subsys.

Signed-off-by: Lang Xie <lang.xie@nordicsemi.no>

### CODE REVIEW

File: samples/mpsl/timeslot/src/main.c [hunk 1]
```diff
@@ -219,26 +238,21 @@ static void mpsl_nonpreemptible_thread(void)
...
-	uint32_t signal_type = 0;
+	uint32_t signal_action = 0;
 
 	while (1) {
-		if (k_msgq_get(&callback_msgq, &signal_type, K_FOREVER) == 0) {
-			switch (signal_type) {
-			case MPSL_TIMESLOT_SIGNAL_START:
+		if (k_msgq_get(&callback_msgq, &signal_action, K_FOREVER) == 0) {
+			switch (signal_action) {
+			case MPSL_TIMESLOT_SIGNAL_ACTION_NONE:
```
Comment (rugeGerritsen) [inline samples/mpsl/timeslot/src/main.c:246]: This looks wrong. `MPSL_TIMESLOT_SIGNAL_ACTION_NONE` should not "Callback: Timeslot start"

File: samples/mpsl/timeslot/src/main.c [hunk 2]
```diff
@@ -120,12 +132,11 @@ static mpsl_timeslot_signal_return_param_t *mpsl_timeslot_callback(
...
-	int err = k_msgq_put(&callback_msgq, &signal_type, K_NO_WAIT);
-
-	if (err) {
-		error();
-	}
+#if defined(NRF5340_XXAA_NETWORK)
+	NVIC_SetPendingIRQ(SWI2_IRQn);
+#elif defined(NRF52840_XXAA) || defined(NRF52832_XXAA) || defined(NRF52832_XXAB)
+	NVIC_SetPendingIRQ(SWI2_EGU2_IRQn);
+#endif
```
Comment (rugeGerritsen) [inline samples/mpsl/timeslot/src/main.c:139]: If you use `CONFIG_SOC_SERIES_NRF52X` or `CONFIG_SOC_SERIES_NRF53X`, the code becomes more future proof. Also, the build should fail if the board is unknown. 

```suggestion
#if defined(NRF5340_XXAA_NETWORK)
	NVIC_SetPendingIRQ(SWI1_IRQn);
#elif defined(CONFIG_SOC_SERIES_NRF52X)
	NVIC_SetPendingIRQ(SWI1_EGU1_IRQn);
#else
#error Unknown device
#endif
```
I would suggest to use `SWI1_EGU1_IRQn` on nRF52 series devices as `SWI2_EGU2_IRQn` does not exist on certain devices in this series.

File: samples/mpsl/timeslot/src/main.c [hunk 3]
```diff
@@ -184,6 +195,14 @@ static void mpsl_nonpreemptible_thread(void)
 	/* Initialize to invalid session id */
 	mpsl_timeslot_session_id_t session_id = 0xFFu;
 
+#if defined(NRF5340_XXAA_NETWORK)
+	IRQ_DIRECT_CONNECT(SWI2_IRQn, 1, swi2_isr, 0);
+	irq_enable(SWI2_IRQn);
+#elif defined(NRF52840_XXAA) || defined(NRF52832_XXAA) || defined(NRF52832_XXAB)
+	IRQ_DIRECT_CONNECT(SWI2_EGU2_IRQn, 1, swi2_isr, 0);
+	irq_enable(SWI2_EGU2_IRQn);
+#endif
```
Comment (rugeGerritsen) [inline samples/mpsl/timeslot/src/main.c:204]: Setting up the IRQs is related to the sample itself, it is not related to setting up the non-preemptible thread. Therefore I would suggest moving it to `main()`

File: samples/mpsl/timeslot/src/main.c [hunk 4]
```diff
@@ -65,6 +65,18 @@ static void error(void)
...
 }
 
+ISR_DIRECT_DECLARE(swi2_isr)
+{
+	/* Put callback info in the message queue. */
+	int err = k_msgq_put(&callback_msgq, &signal_callback_return_param.callback_action,
+						K_NO_WAIT);
+
+	if (err) {
+		error();
```
Comment (rugeGerritsen) [inline samples/mpsl/timeslot/src/main.c:75]: Unrelated to the changes in this PR. I would suggest replacing `error()` with `k_oops()` to ensure we are not using kernel APIs when we can't.

**Review by rugeGerritsen — CHANGES_REQUESTED** [review]
**Timeslot sample**
The approach implemented in this PR introduces a race condition: Multiple high priority signals may be given before the SWI is served. That may cause signals to get lost. That needs to be fixed.

Also, I'm wondering if we should implement a more optimized approach in terms of minimizing the latency from a high priority signal until it is printed in the `console_print_thread`.

The sequence of events in this implementation are as following:
1. High priority signal occurs, SWI is pended. 
2. SWI is served. The signal is put into a message queue.
3. The low priority thread fetches an element from the message queue. 
That is, two intermediate steps are used before the signal is handled/printed. I believe it can be done in one. 

A possible ways of reducing the number intermediate steps:
- Append the signal to a FIFO or ring buffer in the high priority signal callback. This ensures no signals get lost. Note that two FIFOs/ring buffers are needed as some signals are low priority and others are executed from high priority.
- Simply print the signal inside the SWI. By using deferred logging (the default), the logging framework will anyways ensure it is printed in a lower execution context.

**subsys/dm/dm.c**
No changes are needed. The `k_msgq_put(&dm_api_msgq, &dm_api_call, K_NO_WAIT);` was only called from low priority signals, see the docs for `MPSL_TIMESLOT_SIGNAL_SESSION_IDLE`,  `MPSL_TIMESLOT_SIGNAL_BLOCKED`, and `MPSL_TIMESLOT_SIGNAL_CANCELLED`. The return code of this API call should have been checked though.

File: samples/mpsl/timeslot/src/main.c (hunk 1 above)
Comment (laxiLang) [reply samples/mpsl/timeslot/src/main.c:246]: Yes, it is misleading. will fix it.

File: samples/mpsl/timeslot/src/main.c (hunk 4 above)
Comment (laxiLang) [reply samples/mpsl/timeslot/src/main.c:75]: I may misunderstand. But k_oops() is also a kernel function, are you suggesting to define an in-house error function?

**Comment by laxiLang** [comment]
@rugeGerritsen, are you suggesting to define own FIFOs implementation to store messages (to avoid calling kernel queue put and get functions) inside this sample?

**Comment by rugeGerritsen** [comment]
> @rugeGerritsen, are you suggesting to define own FIFOs implementation to store messages (to avoid calling kernel queue put and get functions) inside this sample?

Zephyr provides a FIFO implementation, see the docs

**Comment by laxiLang** [comment]
As required, in the ZLI ISR don’t use the kernel API. If I used only FIFO definition provided by Zephyr kernel service, but define own implementation to write and read it, it seems not simplify too much?

**Comment by rugeGerritsen** [comment]
> As required, in the ZLI ISR don’t use the kernel API. If I used only FIFO definition provided by Zephyr kernel service, but define own implementation to write and read it, it seems not simplify too much?

I assume Zephyr provides a FIFO implementation which is not relying on the kernel itself

File: samples/mpsl/timeslot/src/main.c [hunk 5]
```diff
@@ -50,26 +54,54 @@ static mpsl_timeslot_request_t timeslot_request_normal = {
...
+				LOG_INF("Callback: Session idle\n");
+				break;
+			case MPSL_TIMESLOT_SIGNAL_SESSION_CLOSED:
+				LOG_INF("Callback: Session closed\n");
+				break;
+			default:
+				LOG_INF("Callback: Other signal: %d\n", signal_type);
+				break;
+			}
+		}
```
Comment (rugeGerritsen) [inline samples/mpsl/timeslot/src/main.c:94]: Need to handle the case that there are multiple entries in the ring buffer when the SWI is triggered. Therefore, we need a loop here

File: samples/mpsl/timeslot/src/main.c [hunk 6]
```diff
@@ -87,7 +119,7 @@ static mpsl_timeslot_signal_return_param_t *mpsl_timeslot_callback(
 		nrf_timer_cc_set(NRF_TIMER0, NRF_TIMER_CC_CHANNEL0,
 			TIMER_EXPIRY_US);
 		nrf_timer_int_enable(NRF_TIMER0, NRF_TIMER_INT_COMPARE0_MASK);
-
+		ring_buf_put(&callback_high_priority_ring_buf, &input_data, 1);
```
Comment (rugeGerritsen) [inline samples/mpsl/timeslot/src/main.c:122]: We need to check that the data was actually added

File: samples/mpsl/timeslot/prj.conf [hunk 7]
```diff
@@ -13,3 +13,6 @@ CONFIG_CONSOLE_GETCHAR=y
 CONFIG_MPSL=y
 CONFIG_MPSL_TIMESLOT_SESSION_COUNT=1
 CONFIG_NCS_SAMPLE_EMPTY_APP_CORE_CHILD_IMAGE=y
+CONFIG_RING_BUFFER=y
+CONFIG_LOG=y
+CONFIG_LOG_MODE_DEFERRED=y
```
Comment (rugeGerritsen) [inline samples/mpsl/timeslot/prj.conf:18]: Please add a newline at the end of the file

**Review by rugeGerritsen — CHANGES_REQUESTED** [review]
Looks much better now. Some minor things remaining

**Comment by rugeGerritsen** [comment]
Can you add an entry to the page known_issues.rst?

```
.. rst-class:: v1.9.0 v1-8-0 v1-7-1 v1-7-0 v1-6-1 v1-6-0 v1-5-2 v1-5-1 v1-5-0 v1-4-2 v1-4-1 v1-4-0 v1-3-2 v1-3-1 v1-3-0 v1-2-1 v1-2-0 v1-1-0 v1-0-0

NCSIDB-731: :ref:`timeslot_sample` crashes when calling kernel APIs from zero latency interrupts
  It is not allowed to call kernel APIs from zero latency interrupts.

```

Put the known issue under the section "Bluetooth LE"

File: doc/nrf/known_issues.rst [hunk 8]
```diff
@@ -1024,6 +1024,11 @@ Subsystems
 Bluetooth® LE
```
Comment (wiba-nordic) [inline doc/nrf/known_issues.rst:1024]: Shouldn't this be under MPSL, not Bluetooth?

File: doc/nrf/known_issues.rst [hunk 9]
```diff
@@ -1024,6 +1024,11 @@ Subsystems
...
 =============
 
+.. rst-class:: v1-9-0 v1-8-0 v1-7-1 v1-7-0 v1-6-1 v1-6-0 v1-5-2 v1-5-1 v1-5-0 v1-4-2 v1-4-1 v1-4-0 v1-3-2 v1-3-1 v1-3-0 v1-2-1 v1-2-0 v1-1-0 v1-0-0
+
+NCSIDB-731: :ref:`timeslot_sample` crashes when calling kernel APIs from zero latency interrupts
+  It is not allowed to call kernel APIs from zero latency interrupts.
```
Comment (wiba-nordic) [inline doc/nrf/known_issues.rst:1030]:
```suggestion
  Calling kernel APIs is not allowed from zero latency interrupts.
```

**Review by wiba-nordic — CHANGES_REQUESTED** [review]
A small change, and a question.

**Review by wiba-nordic — APPROVED** [review]

**Pull request merged**